### PR TITLE
ref: Remove `force_final` configuration and simplify `FINAL` handling

### PR DESCRIPTION
**Pull request by tkaemming**
This now has the following behavior, in English:

- If a query has `turbo`, it will never use `FINAL`. These `turbo` queries also do not special case excluded groups to avoid strange phenomena when groups could appear to be "resurrected" if the exclusion set grows larger than the maximum allowable size.
- Queries that are not marked with `turbo` and require final will use `FINAL`.
- Queries that are not marked with `turbo` and do not require final will use the group exclusion set, *unless* the size of the exclusion set exceeds the configured limit, in which case the query will run with `FINAL` instead. (Any queries that use `FINAL` will *not* use the group exclusion set, since the utility of the exclusion set is redundant with the behavior of `FINAL`.)

### CODE REVIEW

**Review by tkaemming — COMMENTED** [review]
This is extracted from some of the work I've been doing on splitting query rewriting and formatting — this probably one of the logically more complicated bits, so figured it'd be good to submit on its own rather than part of the larger PR.

File: snuba/api.py [hunk 1]
```diff
@@ -455,12 +449,10 @@ def error():
     def create_subscription():
         return json.dumps({'subscription_id': uuid1().hex}), 202, {'Content-Type': 'application/json'}
 
-
```
Comment (tkaemming) [inline snuba/api.py]: Oops, will revert these unrelated formatting changes that vscode did.

**Review by fpacifici — COMMENTED** [review]
I don;t see anything wrong here but I am missing context: why do we set "final" if the number of groups to exclude is higher than the maximum ?

**Comment by tkaemming** [comment]
> I don;t see anything wrong here but I am missing context: why do we set "final" if the number of groups to exclude is higher than the maximum ?

This is somewhat described in the description:

> Any queries that use `FINAL` will *not* use the group exclusion set, since the utility of the exclusion set is redundant with the behavior of `FINAL`.

Specifically, the exclusion set is updated when performing a deletion with the groups that were deleted,  or when performing a merge with the groups that were not the destination of the merge.

https://github.com/getsentry/snuba/blob/4b36a882c3ed167dbe84dba9e17eb4b2ec9e5352/snuba/replacer.py#L182
https://github.com/getsentry/snuba/blob/4b36a882c3ed167dbe84dba9e17eb4b2ec9e5352/snuba/replacer.py#L234

Take these examples:

1.  If you delete a group with the id `A` by setting `deleted = 1` on all events where `group_id = A`, then `A` will be added to the exclusion set.
2. If you merge group `A` and `B` into `C` by setting `group_id = C` on all events where `group_id IN (A, B)`, then `A` and `B` would be added to the exclusion set.

The purpose of this set is to allow filtering out the known invalid results that may exist in the dataset until the next `OPTIMIZE` occurs, since adding another expression to the `WHERE` clause should be faster than executing the query with `FINAL`. This works as long as that set of known invalid results is relatively small. 

If the size of that set exceeds the limit, we switch over to using `FINAL` since this will have the same effect as the exclusion set: consolidating all of the records with duplicate IDs.

File: snuba/api.py [hunk 2]
```diff
@@ -216,24 +212,22 @@ def parse_and_run_query(validated_body, timer):
...
-    needs_final, exclude_group_ids = get_projects_query_flags(project_ids)
-    if len(exclude_group_ids) > max_group_ids_exclude:
-        # Cap the number of groups to exclude by query and flip to using FINAL if necessary
-        needs_final = True
-        exclude_group_ids = []
-
-    used_final = False
-    if force_final == 1 or (force_final is None and needs_final):
+    turbo = body.get('turbo', False)
+    if not turbo:
```
Comment (fpacifici) [inline snuba/api.py]: Would it make sense to make this logic dataset specific and imported only by datasets that support merge and deletion ?

**Review by fpacifici — APPROVED** [review]
ok, now i get it

File: snuba/api.py (hunk 2 above)
Comment (tkaemming) [reply snuba/api.py]: Yep, that's my longer-term plan here. 👍Just wanted to get this out of the way first so it's a bit more encapsulated and that change will be a move rather than changing the behavior and moving the code at the same time.
